File: android/src/main/java/com/reactnativemap4dmap/RMFBuildingManager.java
package com.reactnativemap4dmap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;

import java.util.Map;

public class RMFBuildingManager extends ViewGroupManager<RMFBuilding> {

  @NonNull
  @Override
  public String getName() {
    return "RMFBuilding";
  }

  @NonNull
  @Override
  protected RMFBuilding createViewInstance(@NonNull ThemedReactContext reactContext) {
    return new RMFBuilding(reactContext);
  }

  @Override
  public void receiveCommand(@NonNull RMFBuilding view, String commandId, @Nullable ReadableArray args) {
    ReadableMap data;
    switch (commandId) {
      case "setName":
        view.setName(args.getString(0));
        break;
      case "setScale":
        view.setScale(args.getDouble(0));
        break;
      case "setBearing":
        view.setBearing(args.getDouble(0));
        break;
      case "setElevation":
        view.setElevation((float) args.getDouble(0));
        break;
      case "setSelected":
        view.setSelected(args.getBoolean(0));
        break;
      case "setVisible":
        view.setVisible(args.getBoolean(0));
        break;
      case "setTouchable":
        view.setTouchable(args.getBoolean(0));
        break;
      case "setUserData":
        data = args.getMap(0);
        view.setUserData(data);
        break;
    }
  }

  @Override
  public Map getExportedCustomDirectEventTypeConstants() {
    final Map<String, Map<String, String>> map = MapBuilder.of("onPress", MapBuilder.of("registrationName", "onPress"));
    return map;
  }

  @ReactProp(name = "name")
  public void setName(final RMFBuilding view, final String name) {
    view.setName(name);
  }

  @ReactProp(name = "coordinate")
  public void setCoordinate(final RMFBuilding view, final ReadableMap map) {
    view.setCoordinate(map);
  }

  @ReactProp(name = "modelUrl")
  public void setModelUrl(final RMFBuilding view, final String url) {
    view.setModelUrl(url);
  }

  @ReactProp(name = "textureUrl")
  public void setTextureUrl(final RMFBuilding view, final String url) {
    view.setTextureUrl(url);
  }

  @ReactProp(name = "scale")
  public void setScale(final RMFBuilding view, final double scale) {
    view.setScale(scale);
  }

  @ReactProp(name = "bearing")
  public void setBearing(final RMFBuilding view, final double bearing) {
    view.setBearing(bearing);
  }

  @ReactProp(name = "elevation")
  public void setElevation(final RMFBuilding view, final float elevation) {
    view.setElevation(elevation);
  }

  @ReactProp(name = "selected")
  public void setSelected(final RMFBuilding view, final boolean selected) {
    view.setSelected(selected);
  }

  @ReactProp(name = "visible")
  public void setVisible(final RMFBuilding view, final boolean visible) {
    view.setVisible(visible);
  }

  @ReactProp(name = "touchable")
  public void setTouchable(final RMFBuilding view, final boolean touchable) {
    view.setTouchable(touchable);
  }

  @ReactProp(name = "userData")
  public void setUserData(final RMFBuilding view, final ReadableMap userData) {
    view.setUserData(userData);
  }
}

File: ios/building/RMFBuildingManager.m
//
//  RMFBuildingManager.m
//  react-native-map4d-map
//
//  Created by Huy Dang on 6/10/25.
//  Copyright © 2025 IOTLink. All rights reserved.
//

#import "RMFBuildingManager.h"
#import "RMFBuilding.h"
#import <React/RCTConvert+CoreLocation.h>
#import <React/RCTBridge.h>
#import <React/RCTUIManager.h>

@implementation RMFBuildingManager

RCT_EXPORT_MODULE(RMFBuilding)

- (void)withBuildingForTag:(nonnull NSNumber *)reactTag
                   handler:(void (^)(RMFBuilding *building))handler
{
  RCTUIManager *uiManager = self.bridge.uiManager;
  if (uiManager == nil) {
    RCTLogError(@"UIManager is unavailable on iOS runtime");
    return;
  }

  [uiManager addUIBlock:^(__unused RCTUIManager *manager, NSDictionary<NSNumber *, UIView *> *viewRegistry) {
    id view = viewRegistry[reactTag];
    if (![view isKindOfClass:[RMFBuilding class]]) {
      RCTLogError(@"Invalid view returned from registry, expecting RMFBuilding, got: %@", view);
      return;
    }

    handler((RMFBuilding *)view);
  }];
}

- (UIView *)view {
  RMFBuilding *building = [[RMFBuilding alloc] init];
  return building;
}

RCT_EXPORT_VIEW_PROPERTY(name, NSString)
RCT_EXPORT_VIEW_PROPERTY(coordinate, CLLocationCoordinate2D)
RCT_EXPORT_VIEW_PROPERTY(modelUrl, NSString)
RCT_EXPORT_VIEW_PROPERTY(textureUrl, NSString)
RCT_EXPORT_VIEW_PROPERTY(scale, double)
RCT_EXPORT_VIEW_PROPERTY(bearing, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(elevation, double)
RCT_EXPORT_VIEW_PROPERTY(selected, BOOL)
RCT_EXPORT_VIEW_PROPERTY(touchable, BOOL)
RCT_EXPORT_VIEW_PROPERTY(visible, BOOL)
RCT_EXPORT_VIEW_PROPERTY(userData, NSDictionary)
RCT_EXPORT_VIEW_PROPERTY(onPress, RCTBubblingEventBlock)

RCT_EXPORT_METHOD(setName:(nonnull NSNumber *)reactTag name:(NSString *)name) {
  [self withBuildingForTag:reactTag handler:^(RMFBuilding *building) {
    [building setName:name];
  }];
}

RCT_EXPORT_METHOD(setScale:(nonnull NSNumber *)reactTag scale:(double)scale) {
  [self withBuildingForTag:reactTag handler:^(RMFBuilding *building) {
    [building setScale:scale];
  }];
}

RCT_EXPORT_METHOD(setBearing:(nonnull NSNumber *)reactTag bearing:(CGFloat)bearing) {
  [self withBuildingForTag:reactTag handler:^(RMFBuilding *building) {
    [building setBearing:bearing];
  }];
}

RCT_EXPORT_METHOD(setElevation:(nonnull NSNumber *)reactTag elevation:(double)elevation) {
  [self withBuildingForTag:reactTag handler:^(RMFBuilding *building) {
    [building setElevation:elevation];
  }];
}

RCT_EXPORT_METHOD(setSelected:(nonnull NSNumber *)reactTag selected:(BOOL)selected) {
  [self withBuildingForTag:reactTag handler:^(RMFBuilding *building) {
    [building setSelected:selected];
  }];
}

RCT_EXPORT_METHOD(setTouchable:(nonnull NSNumber *)reactTag touchable:(BOOL)touchable) {
  [self withBuildingForTag:reactTag handler:^(RMFBuilding *building) {
    [building setTouchable:touchable];
  }];
}

RCT_EXPORT_METHOD(setVisible:(nonnull NSNumber *)reactTag visible:(BOOL)visible) {
  [self withBuildingForTag:reactTag handler:^(RMFBuilding *building) {
    [building setVisible:visible];
  }];
}

RCT_EXPORT_METHOD(setUserData:(nonnull NSNumber *)reactTag userData:(id)json) {
  [self withBuildingForTag:reactTag handler:^(RMFBuilding *building) {
    [building setUserData:[RCTConvert NSDictionary:json]];
  }];
}

@end
